runeq/resources/project.py

```diff
@@ -9,6 +9,311 @@
 from .common import ItemBase, ItemSet
 
 
+class Metric(ItemBase):
+    """
+    A measurement related to processed patient data. Each metric is related
+    to a single processed data type, project, and time interval. These metrics
+    are calculated periodically and updated.
+
+    """
+    def __init__(
+        self,
+        id: str,
+        type: str,
+        data_type: str,
+        time_interval: str,
+        updated_at: float,
+        created_at: float,
+        value: Optional[float] = None,
+    ):
+        """
+        Initialize with data.
+
+        Args:
+            id: Unique identifier of the patient
+            type: Type of measurement/metric. Possible types include:
+                TOTAL_HOURS, LATEST_DATA, LAST_UPLOAD
+            data_type: Processed stream data type that is being measured.
+                Possible data types include:
+                    APPLEWATCH_SYMPTOM, APPLEWATCH_TREMOR,
+                    APPLEWATCH_DYSKINESIA, APPLEWATCH_HEART_RATE,
+                    PERCEPT_TREND_LOG_LFP
+            value:
+            time_interval: Period over which the metric was calculated.
+                Possible time intervals include:
+                    FOURTEEN_DAYS, NINETY_DAYS, PROJECT_ALL
+            created_at: Time patient was added to the cohort (unix timestamp)
+            updated_at: Time cohort patient was last updated (unix timestamp)
+
+        """
+        self.type = type
+        self.data_type = data_type
+        self.time_interval = time_interval
+        self.created_at = created_at
+        self.updated_at = updated_at
+        self.value = value
+
+        super().__init__(
+            id=id,
+            type=type,
+            data_type=data_type,
+            value=value,
+            time_interval=time_interval,
+            created_at=created_at,
+            updated_at=updated_at,
+        )
+
+
+class MetricSet(ItemSet):
+    """
+    A collection of Metrics.
+
+    """
+
+    def __init__(self, items: Iterable[Metric] = ()):
+        """
+        Initialize with Metrics.
+
+        """
+        super().__init__(items=items)
+
+    @property
+    def _item_class(self) -> Type[ItemBase]:
+        """
+        Instance type of items in this set.
+
+        """
+        return Metric
+
+
+class CohortPatient(ItemBase):
```

Comment:
It looks like there is no implementation difference between a `CohortPatient` and a `ProjectPatient`. I also think it might be pretty complex for an SDK user to keep track of regular Patients, CohortPatients and ProjectPatients. Also the `CohortPatient` and `ProjectPatient` appear to have no devices attached to them, so this would mean an SDK user would have to separately create a `Patient` (who has devices) and a corresponding `CohortPatient` and a corresponding `ProjectPatient` to keep track of the same patient.

- What are your thoughts on removing these specific instances of `CohortPatient` and `ProjectPatient`, and instead adding the `MetricSet` as an attribute on the regular `Patient` class. The `Patient` class already supports `**attributes`, so all you would have to do when calling `get_project_patients()` or `get_cohort_patients()` is return a `PatientSet` where each `Patient` is instantiated with a `metric` field as one of it's attributes.

- Another option is to inherit from `Patient` as a base class here instead of `ItemBase`. This abstraction pattern should minimize your code here, and make it easier for SDK users to work with.

Both suggestions would involve adding to the graphql query to fetch patient devices. Or you could call `get_patient` within the `get_project_patients` and `get_cohort_patients` classes to fetch that data for you, and return one patient with all their data together.

These suggestions also apply to `CohortPatientSet`, `ProjectPatientSet`, `Cohort` and `CohortSet`.

@cranti also interested on your thoughts here

Reply:
So while they are indeed similar, I do think it is important to keep them separate because they are associated with different objects (project or cohort). This can be made clearer by adding the project or cohort id to each item. I don't think there is any large value in adding additional graphql queries. I can make a base class that they inherit from, which makes sense. 

You are correct that devices are not associated with them.

Reply:
I agree with @abhinavpottabathula -- it seems confusing to have so many different implementations of "patient". I'm also not really following the utility of the project/cohort distinction in these classes.

That said, I'm not quite familiar enough with projects & cohorts to fully weigh in -- maybe the three of us could hop on a quick call next week? It would be helpful to see some example code, for how you anticipate this being used.

Reply:
Sure, we can hop on a call. It would be good to have @gavin-philips as well since he has been providing input for what the SDK functionality should be for projects.